Report/sections/2_design/expressions.tex

```diff
@@ -1 +1,37 @@
+Arithmetic expressions are basic mathematical expressions.
+They consist of operators, operands, and function calls. 
+Operators can either be \textit{unary}, \textit{binary} or \textit{ternary} meaning they have either one, two or three operands.
+Finally, binary operators are usually \textit{infix}, meaning they appear between operands\cite{sebesta_concepts_2016}.
+
+\subsubsection*{Operator evaluation order}
+The \textit{precedence} of an operator along with the \textit{associativity} rules in a language dictate the order in which operators are evaluated.
+Different languages implement this differently. Some languages simply evaluate operators in the order that they appear. Usually, however, languages follow the order of operations defined in mathematics such as multiplication being of higher priority than addition\cite{sebesta_concepts_2016}.
+
+
+Parentheses are also often part of expressions in programming languages and typically impacts the evaluation order of operators such that the contents of the parentheses are evaluated first\cite{sebesta_concepts_2016}.
+
+
+Associativity in languages is most commonly implemented as left to right, however some operators like the exponentation operator sometimes associates right to left\cite{sebesta_concepts_2016}. 
+
+\subsubsection*{Relational expressions}
+A relational expression is an expression that has two operands and one relational operator. The relational operator compares the values of the two operands and returns a
+boolean, if booleans are included in the language. Relational operators include symbols such as \texttt{LESS THAN} and \texttt{LARGER THAN}, as well as \texttt{EQUAL TO} and
+\texttt{NOT EQUAL TO}. 
+Relational operators will typically have lower precedence than arithmetic operators to ensure proper evaluation\cite{sebesta_concepts_2016}.
+
+
+\subsubsection*{Boolean expressions}
+Boolean expressions are constructed using different mathematical logic operators such as the logical \texttt{AND}, \texttt{OR} and \texttt{EQUALS}. 
+When evaluated, a whole expression is either true or false. 
+This is then often combined with statements such as \texttt{if-then-else} statements to achieve a conditional program flow - if the expression evaluates to true, the code in the \texttt{then} block is executed, and if it evaluates to false, code in the \texttt{else} block is executed instead\cite{sebesta_concepts_2016}.
+
+\subsubsection*{Expressions in \dazel{}}
+\dazel{} features a variety of binary operators to support arithmetic as well as unary operations for boolean expressions.
+It also has an operator evalation order identical to mathematics, as well as a left to right associativity. This helps prevent confusion for beginner programmers, 
+and also maintains the standard that most experienced programmers are accustomed to.
+To accomodate evalation of expressions \dazel{} also has the most common relational and logic operators as can be seen in \tabref{table:operatorsymbols}. 
```

Comment:
This entire section reads as if it were written off from Sebasta - or, at least, someone's notes on the book. The introduction is weird and does not say anything. We don't really deliver on the promise to explain the relevance of the types to Dazel - we just explain what the types are.

There are too many subsections.

My suggestion: reduce this section to the very last few lines. Expand upon those. Line 34 mentions a 'variety of binary operators'. Refer to the table here. Show what Dazel has.

```suggestion
\dazel{} supports a variety of binary operators for arithmetic expressions, as well as unary operations for boolean expressions. These operators can be seen in \tabref{table:operatorsymbols}.
Operator evalation order in \dazel{} is identical to that seen in mathematics.
```

We have no data to back up the claim that "_This helps prevent confusion for beginner programmers, 
and also maintains the standard that most experienced programmers are accustomed to._" - unless, of course, I've missed something. In that case, please cite it here.

Perhaps move the table containing the operators (and the section that introduces it) closer to this section. Or merge them.

Reply:
We need to introduce all the important concepts and theory before we can argue why Dazel works the way it does. Similarly to other sections, such as the one about scopes, we first describe the concepts, then relate them to Dazel, as was requested by the supervisor.